mirgecom/steppers.py

```diff
@@ -157,14 +177,24 @@ def _advance_state_leap(rhs, timestepper, checkpoint, get_timestep,
         if dt < 0:
             return istep, t, state
 
-        checkpoint(state=state, step=istep, t=t, dt=dt)
+        if pre_step_callback is not None:
+            state = pre_step_callback(state=state,
+                                      step=istep,
+                                      t=t, dt=dt)
 
         # Leap interface here is *a bit* different.
         for event in stepper_cls.run(t_end=t+dt):
             if isinstance(event, stepper_cls.StateComputed):
                 state = event.state_component
                 t += dt
+
+                if post_step_callback is not None:
+                    state = post_step_callback(state=state,
```

Comment:
This doesn't work to modify state. Leap maintains its own state, and it has its own callbacks to modify state. IMO state modification should be kept separate from this, it's a concern mainly for the time integrator. A "time step" is not a well-defined interval anyhow, cf. multi-step vs multi-state methods.

Reply:
We don't have a current use for this, but the advancer calls "tick after", etc for the logging.  If I had my druthers, we'd evict logger from the stepper to the driver level and let it get called as a part of this pre/post step callback infrastructure, or become part of whatever automatic library routines we roll for these utilities.

Reply:
See (100df77) `autoignition-mpi.py` and `vortex-mpi.py` for a sample of what it looks like to do the log functions in callback.

Reply:
> This doesn't work to modify state. Leap maintains its own state, and it has its own callbacks to modify state. IMO state modification should be kept separate from this, it's a concern mainly for the time integrator.

@inducer Could you clarify what you mean here? I had thought we sorted this out in a previous PR discussion w.r.t. filtering/limiting.

Reply:
Two things:
* The intervals at which the time stepper returns to the user are not physically meaningful. (So: doing something (like filtering/limiting) every time that happens is also not a good idea.) For example, multi-step methods do one RHS evaluation per step, and their steps are correspondingly smaller than those of multi-stage (e.g. RK) methods. Multi-rate methods further upend this.
* The second thing is mostly a software concern to do with leap: It assumes that it is the sole owner of ODE state. If you'd like to know the state, you ask it. If you'd like to modify it, you do so in a callback coming from its guts. 

My conclusion from this is that we shouldn't pursue state modification in a way that isn't driven by the time integrator.

Reply:
FWIW, in our current code the state is returned to the user after every single major step.  I feel like the user should own the state between time steps.  I still don't understand the concern.  

After:

`next_state = perform_timestep(current_state, ...)`

Isn't `next_state` "mine"?  Can't I do whatever I want to it?  Whatever `perform_timestep` does on the inside to produce that `next_state` is the `perform_timestep`'s business and we are not trying to modify the state down inside that, only after the stepper has finished producing the next state.

Reply:
> Isn't next_state "mine"?

IMO: It really isn't.

If the time stepper keeps history (like multi-step ones do), then any state modification will be inconistent with the history. That's just mathematically true. What you get if you do that isn't really known, but I'd guess an order drop in temporal accuracy at the least, unless you re-bootstrap the multi-step methods, which is slow and expensive.

As a software matter, leap models time integration as a [coroutine](https://en.wikipedia.org/wiki/Coroutine). You *can* inject new state at any time that control returns to you, but that's about as elegant as someone "from the outside" poking variable values into your namespace.

Lastly, as a combined math/CS concern: The points in time when state modification should happen should really be signalled by the time integrator. "Once every time step" means something different for Runge-Kutta than for Adams-Bashforth.

Reply:
Ok, then I think this will be my last attempt to try and motivate the retention of the state modification ability by the user callback.   

I follow the argument about time steppers that keep a history - but does this justify removing the ability to modify the state altogether?  We aren't using those steppers, and even if we ever did we'd know to turn off any filtering or state modification.  I guess I am of the opinion that even though sometimes it is not a good idea to do it, that we shouldn't forbid (or disable) them from doing so by changing the return signature.

We actually have test cases that use spectral filtering, and they filter the solution between time steps, and iirc, this was the recommended strategy.  This case *could* use the library built-in stepper with the proposed API, but if we hard prevent state modification between steps, then this test case is relegated to using a custom "test" stepper that allows it.   

I can see a user wanting to filter the state, or limit it or twiddle some sort of actuators in it, and even though it is not always a good idea to do so, I don't think hard preventing it through return signature restriction is the greatest idea.

Reply:
It's probably not worth it to have this discussion preemptively, let's wait until the issue arises.